### PR TITLE
Support custom MicroProfile Config converters

### PR DESCRIPTION
**Pull request by sarxos**
Hello @dmlloyd, 

This is a commit I prepared in regards to the problem we discussed on Zulip to address issue #702 you created in the past.

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (dmlloyd) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: The SmallRye config implementation uses a different approach to determine the converter type:

```java
    private Type getConverterType(Class clazz) {
        if (clazz.equals(Object.class)) {
            return null;
        }

        for (Type type : clazz.getGenericInterfaces()) {
            if (type instanceof ParameterizedType) {
                ParameterizedType pt = (ParameterizedType) type;
                if (pt.getRawType().equals(Converter.class)) {
                    Type[] typeArguments = pt.getActualTypeArguments();
                    if (typeArguments.length != 1) {
                        throw new IllegalStateException("Converter " + clazz + " must be parameterized with a single type");
                    }
                    return typeArguments[0];
                }
            }
        }

        return getConverterType(clazz.getSuperclass());
    }
```

Granted they're using raw types in there (unnecessarily), but it's probably worth ensuring that we're doing the same process.  We could even use this exact method (reflectively at first using an approach similar to that employed by `io.quarkus.runtime.configuration.ConverterFactory#getImplicitConverter`, and then later by making that SmallRye Config method be publicly accessible somehow).

**Review by dmlloyd — CHANGES_REQUESTED** [review]
A few notes.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Hi @dmlloyd, I used `getConverterType` method as requested and put reflection code in the `ConverterFactory` since it looked to be a good place for a dirty code. After changes here are approved I will prepare a PR to SmallRye repo to make `getConverterType` public static or probably move it to https://github.com/smallrye/smallrye-config/blob/70abffdaadbdcf21ddf5c976157c517e29735f8d/implementation/src/main/java/io/smallrye/config/Converters.java or https://github.com/smallrye/smallrye-config/blob/70abffdaadbdcf21ddf5c976157c517e29735f8d/implementation/src/main/java/io/smallrye/config/ImplicitConverters.java class since these seems to be a better for this kind of functionality.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigurationCustomConverterBuildItem.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigurationCustomConverterBuildItem.java]: Yup, it is possible. Done.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: @dmlloyd Thank you for inspecting this code. Before we move forward I would like to discuss some caveats I see in regards to this idea. I may be wrong since I'm not very familiar with the SubstrateVM, but after looking at [this code](https://github.com/cstancu/native-image-service-loader-demo), specifically this [reflections config](https://github.com/cstancu/native-image-service-loader-demo/blob/master/reflection_config.json), I assume that there is a need to produce `ReflectiveClassBuildItem` for every `Converter` implementation we have in our services file. This in turn must be done in `@BuildStep` thus leading us back to the deployment time `Converter` iteration. Therefore it seems impossible to me to move this to the runtime because even when we do it, we still have to have `ReflectiveClassBuildItem` produced in deployment time, which is charged to the same problem you pointed out in your comment.

Comment (dmlloyd) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: This is not exactly correct.  Reflection actually works completely fine in native images, as long as you have the reflection object during static initialization.  The reflective class build items are not as magical as they seem: really all they do is create the reflection objects during static initialization so that they're available at run time.  By creating the `Converter` instances during static initialization, even the ones which use reflection, we basically bypass this entire process and no reflective build items are needed.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (dmlloyd) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: This is already done: https://github.com/smallrye/smallrye-config/pull/111

Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Hah! Nice :)

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Ok, I get it now. So if reflections are allowed during static initialization, my idea is the following.

Modify synthetic `io/quarkus/runtime/generated/RunTimeConfig.class` by adding `static final` field whitch will hold a list of all `Converter` impls, something like:

```java
private static final List<ConverterItem<?>> CONVERTERS = ConverterSupport.getConverters();
```

And have a `ConverterSupport` class like this:

```java
public class ConverterSupport {

  static final int DEFAULT_CONVERTER_PRIORITY = 200;

  static int getConverterPriority(final Class<? extends Converter<?>> converterClass) {
    return Optional
      .ofNullable(converterClass.getAnnotation(Priority.class))
      .map(Priority::value)
      .orElse(DEFAULT_CONVERTER_PRIORITY);
  }

  private static <T> ConverterItem<?> converterToItem(final Converter<T> converter) {
    final Class<? extends Converter<T>> converterClass = (Class<? extends Converter<T>>) converter.getClass();
    final Class<T> convertedType = ConverterFactory.getConverterType(converter);
    final int priority = getConverterPriority(converterClass);
    return new ConverterItem<T>(convertedType, converter, priority);
  }

  public static List<ConverterItem<?>> getConverters() {
    return StreamSupport
      .stream(ServiceLoader.load(Converter.class).spliterator(), false)
      .map(ConverterSupport::converterToItem)
      .collect(Collectors.toList());
  }
}
```

And modify synthetic `io/quarkus/runtime/generated/RunTimeConfig.class` again to change `getRunTimeConfiguration` method by adding `for/each` loop which will iterate over items in `CONVERTERS` list and populate `SmallRyeConfigBuilder`:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  for (ConverterItem item : CONVERTERS) {
    builder.withConverter(item.getConvertedType(), item.getPriority(), item.getConverter());
  }
  // some other stuff
}
```

Instead of:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  builder.withConverter(SomeType1.class, 200, new SomeTypeConverter1());
  builder.withConverter(SomeType2.class, 200, new SomeTypeConverter2());
  builder.withConverter(SomeType3.class, 200, new SomeTypeConverter3());
  builder.withConverter(SomeType4.class, 200, new SomeTypeConverter4());
  // some other stuff
}
```

But I'm completely green when it comes to Gizmo :( I have no idea where to start, i.e. how to generate a basic foor loop. I once asked about some Gizmo related materials (tutorials, demos, examples, etc) on Zulip but since there was no response I assumed that such goodies are not available.

Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Ok, I have another idea. Much simpler and doable with my level of Gizmo experience :) The `ConverterSupport` class will look like this:

```java
public class ConverterSupport {

  private static final List<ConverterItem<?>> CONVERTERS = getConverters();
  private static final int DEFAULT_CONVERTER_PRIORITY = 200;

  static int getConverterPriority(final Class<? extends Converter<?>> converterClass) {
    return Optional
      .ofNullable(converterClass.getAnnotation(Priority.class))
      .map(Priority::value)
      .orElse(DEFAULT_CONVERTER_PRIORITY);
  }

  private static <T> ConverterItem<?> converterToItem(final Converter<T> converter) {
    final Class<? extends Converter<T>> converterClass = (Class<? extends Converter<T>>) converter.getClass();
    final Class<T> convertedType = ConverterFactory.getConverterType(converter);
    final int priority = getConverterPriority(converterClass);
    return new ConverterItem<T>(convertedType, converter, priority);
  }

  private static List<ConverterItem<?>> getConverters() {
    return StreamSupport
      .stream(ServiceLoader.load(Converter.class).spliterator(), false)
      .map(ConverterSupport::converterToItem)
      .collect(Collectors.toList());
  }

  public static void populate(final SmallRyeConfigBuilder builder) {
    CONVERTERS.forEach(addConverterTo(builder));
  }

  private static <T> Consumer<ConverterItem<T>> addConverterTo(final SmallRyeConfigBuilder builder) {
    return item -> builder.withConverter(
      item.getConvertedType(),
      item.getPriority(),
      item.getConverter());
  }
}
```

And in `io/quarkus/runtime/generated/RunTimeConfig.class` it will be used as:

```java
public static void getRunTimeConfiguration() {
  SmallRyeConfigBuilder builder = new SmallRyeConfigBuilder();
  // some stuff
  ConverterSupport.populate(builder);
  // some other stuff
}
```

I think invoking single static method with one argument is simpler than generating `for/each` loop. 

What do you think?

Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Additionally we can also have a `@Template` which will be invoked in a `@Record(STATIC_INIT)` `@BuildStep` and initialize the `ConverterSupport` class.

Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: @dmlloyd, Can you please comment?

Comment (dmlloyd) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: It seems reasonable.  You don't have to initialize the `ConverterSupport` class because the `native-image` tool initializes all classes for you.

Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Ok, this is a good news :) I will start modifying the code and will come back to you soon.

File: core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java
Comment (sarxos) [inline core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java]: This was removed because this converter is now included in services file and therefore there is no need to produce `ConfigurationCustomConverterBuildItem`.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigurationCustomConverterBuildItem.java
Comment (sarxos) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigurationCustomConverterBuildItem.java]: There is no need for this class any more.

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: I'm not a big fan of this pattern.

AFAICS, you don't need to do that as the `populateConverters` method is called in a `@BuildStep` so it's called when building the jar.

You could simply initialize them when in `populateConverters`. Or do I miss something?

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: Could you please get rid of all the `@author` tags in the PR? We decided we do not want them.

File: core/runtime/src/test/java/io/quarkus/runtime/configuration/ConverterSupportTest.java
Comment (gsmet) [inline core/runtime/src/test/java/io/quarkus/runtime/configuration/ConverterSupportTest.java]: s/tests cases/test case/

**Review by gsmet — COMMENTED** [review]
I'll let @dmlloyd review it in depth but I added a few cosmetic comments.

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (sarxos) [reply core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: Sure. My eclipse is generates them automatically for all of my classes.

File: core/runtime/src/test/java/io/quarkus/runtime/configuration/ConverterSupportTest.java
Comment (sarxos) [reply core/runtime/src/test/java/io/quarkus/runtime/configuration/ConverterSupportTest.java]: Done.

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (sarxos) [reply core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: Hi @gsmet. In regards to your comment:

1. Can you please precise which patter did you refer to? Generic `Object` or static initialization of `ConverterItem` `List`?
2. These converter's constructors must be available for the reflection because `ServiceLoader` won't be able to initialize them and since we cannot produce reflective class build items, we need to make sure to have these converters available during static initialization because when reflected object is available in static initialization there is no need to configure reflection (this is what I learned from @dmlloyd). Loading them in `cinit` guarantees that converters are initialized because `native-image` process and initializes every class. If initialize this `ConverterItem` `List` in method body then it can only be initialized indirectly, when invocation to `populateConverters(builder)` is done from a static context of other class, which is not guaranteed at all. It's will be only assumed.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Hi @dmlloyd 

Please take a look again. Some changes worth noting:

* `ConfigurationCustomConverterBuildItem` is no more. With new mechanism it became obsolete and I removed it.
* Updated documentation.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (dmlloyd) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Please avoid changes not related to the issue as it makes it hard to review.

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: The converter list has to be in a static field in order to make it available to the native image run time.  In general I agree with this approach.

**Review by dmlloyd — CHANGES_REQUESTED** [review]
I think after the priorities are added to our converters, this will be ready to go.

File: core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java
Comment (gsmet) [reply core/runtime/src/main/java/io/quarkus/runtime/configuration/ConverterSupport.java]: Ah yes, sorry, I missed we did call it again at a later stage.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: @dmlloyd Ok. I will change `DEFAULT_CONVERTER_PRIORITY` in `ConverterSupport`  to be 100 instead of 200 and add `@Priority` to all Quarkus explicit converters.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: Sure, but I will leave these javadocs I added here. I noticed that the number of params didn't match arguments count here so I added missing ones. Indeed, Github's code review is not very readable.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
Comment (sarxos) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java]: @dmlloyd This is now done. I added `@Priority(DEFAULT_QUARKUS_CONVERTER_PRIORITY)`, on all Quarkus converters, which has a value of 200. All converters made by users will have a default priority of 100.

**Comment by sarxos** [comment]
Hi @dmlloyd, @gsmet New patchset is ready for next code review iteration. All prior comments are addressed.

**Review by dmlloyd — APPROVED** [review]

**Comment by dmlloyd** [comment]
Looks OK, thanks.

**Pull request merged**
